src/cim/components/Agent/tableUtils.tsx

```diff
@@ -222,38 +227,118 @@ export const infraEnvColumn = (agents: AgentK8sResource[]): TableRow<Host> => {
 };
 
-export const canEditBMH = (bmh: BareMetalHostK8sResource) =>
-  [
+export const canEditBMH = (bmh: BareMetalHostK8sResource): ActionCheck => {
+  const canEdit = [
     bmhStatus.provisioned.key,
     bmhStatus.deprovisioning.key,
     bmhStatus.pending.key,
     bmhStatus.registering.key,
-  ].includes(getBMHStatus(bmh).state.key)
-    ? 'Bare metal host hostname cannot be changed once configured. Remove this host and discover it again if a change is needed.'
-    : undefined;
+  ].includes(getBMHStatus(bmh).state.key);
+  return [
+    canEdit,
+    canEdit
+      ? undefined
+      : 'Bare metal host cannot be edited. Remove this host and add it again if a change is needed.',
+  ];
+};
+
+export const canEditAgent = (agent: AgentK8sResource): ActionCheck => {
+  const enabled = getAgentStatus(agent).status.category !== 'Installation related';
+  return [
+    enabled,
+    enabled
+      ? undefined
+      : 'Hostname cannot be edited while host is either installed or being installed.',
+  ];
+};
+
+export const canChangeHostname = (
+  agents: AgentK8sResource[],
+  bareMetalHosts: BareMetalHostK8sResource[],
+): ((h: Host) => ActionCheck) => (h: Host) => {
+  const agent = agents.find((a) => a.metadata?.uid === h.id);
+  if (agent) {
+    return canEditAgent(agent);
+  }
+  const bmh = bareMetalHosts.find((bmh) => bmh.metadata?.uid === h.id);
+  if (bmh) {
+    return canEditBMH(bmh);
+  }
+  return [true, undefined];
+};
+
+export const canUnbindAgent = (
+  agentClusterInstalls: AgentClusterInstallK8sResource[] | undefined,
+  agent: AgentK8sResource,
+): ActionCheck => {
+  if (!agent?.spec.clusterDeploymentName?.name) {
+    return [false, 'The agent is not bound to a cluster.'];
+  }
+
+  const { status } = getAgentStatus(agent);
+  if (
+    [
+      'preparing-for-installation',
+      'installing',
+      'installing-in-progress',
+      'installing-pending-user-action',
+      'resetting-pending-user-action',
+    ].includes(status.key)
+  ) {
+    return [false, 'It is not possible to remove a host which is being installed.'];
+  }
+
+  if (
+    ['installed', 'error', 'cancelled'].includes(status.key) &&
+    (agent.status?.role === 'master' || agent.status?.role === 'bootstrap')
+  ) {
+    return [false, 'It is not possible to remove control plane node from an installed cluster.'];
+  }
+
+  if (agentClusterInstalls) {
+    const agentClusterInstall = getAgentClusterInstallOfAgent(agentClusterInstalls, agent);
+
+    if (agentClusterInstall) {
+      if (isInstallationInProgress(agentClusterInstall)) {
+        return [false, 'It is not possible to remove a node from a cluster during installation.'];
+      }
+
+      if (
+        ['installed', 'error', 'cancelled'].includes(status.key) &&
+        agent.status?.role === 'worker' &&
+        agentClusterInstall.spec?.provisionRequirements?.workerAgents === 2
+      ) {
+        return [false, 'This host cannot be removed as there are only 2 workers in the cluster.'];
```

Comment:
IIUC this will prevent the user from removing workers if there are 2 of them. What if user has 3 workers and removes them? Will we only allow to remove 1 of them? If not, should we enforce this when there are only 2 workers?

I am inclined towards allowing user to remove workers from the cluster. Enforcing various cases on the UI side can lead to bugs. Backend does not enforce it AFAIK.

Reply:
removed